### PR TITLE
Add limits for silences

### PR DESCRIPTION
**Pull request by grobinson-grafana**
This commit adds limits for silences including the maximum number of active and pending silences, and the maximum size per silence (in bytes).

### CODE REVIEW

File: silence/silence.go [hunk 1]
```diff
@@ -611,7 +645,11 @@ func (s *Silences) Set(sil *pb.Silence) (string, error) {
 		sil.StartsAt = now
 	}
 
-	return sil.Id, s.setSilence(sil, now, false)
```
Comment (grobinson-grafana) [inline silence/silence.go:614]: We shouldn't return the ID when no silence is created.

File: silence/silence.go [hunk 2]
```diff
@@ -755,6 +793,9 @@ func (s *Silences) QueryOne(params ...QueryParam) (*pb.Silence, error) {
 // Query for silences based on the given query parameters. It returns the
 // resulting silences and the state version the result is based on.
 func (s *Silences) Query(params ...QueryParam) ([]*pb.Silence, int, error) {
+	s.mtx.Lock()
```
Comment (grobinson-grafana) [inline silence/silence.go:796]: Moved the mutex from `query` to `Query` so we can call `query` from `Set` while the mutex is held.

File: silence/silence.go [hunk 3]
```diff
@@ -600,7 +618,24 @@ func (s *Silences) Set(sil *pb.Silence) (string, error) {
...
+		err := QState(types.SilenceStateActive, types.SilenceStatePending)(q)
+		if err != nil {
+			return "", err
+		}
+		sils, _, err := s.query(q, s.nowUTC())
+		if err != nil {
+			return "", err
+		}
+		if len(sils)+1 > s.limits.MaxSilences {
+			return "", fmt.Errorf("exceeded maximum number of silences: %d", s.limits.MaxSilences)
```
Comment (gotjosh) [inline silence/silence.go:635]:
```suggestion
			return "", fmt.Errorf("exceeded maximum number of silences: %d (limit: %d)", len(sils), s.limits.MaxSilences)
```

File: silence/silence.go [hunk 4]
```diff
@@ -600,7 +618,24 @@ func (s *Silences) Set(sil *pb.Silence) (string, error) {
...
+	if s.limits.MaxSilences > 0 {
+		// Get the number of active and pending silences to enforce limits.
+		q := &query{}
+		err := QState(types.SilenceStateActive, types.SilenceStatePending)(q)
+		if err != nil {
+			return "", err
+		}
+		sils, _, err := s.query(q, s.nowUTC())
+		if err != nil {
+			return "", err
```
Comment (gotjosh) [inline silence/silence.go:632]:
```suggestion
			return "", fmt.Errorf("unable to query silences while calculating the limit: %w", err)
```

File: silence/silence.go [hunk 5]
```diff
@@ -600,7 +618,24 @@ func (s *Silences) Set(sil *pb.Silence) (string, error) {
...
 		}
 	}
+
 	// If we got here it's either a new silence or a replacing one.
+	if s.limits.MaxSilences > 0 {
+		// Get the number of active and pending silences to enforce limits.
+		q := &query{}
+		err := QState(types.SilenceStateActive, types.SilenceStatePending)(q)
+		if err != nil {
+			return "", err
```
Comment (gotjosh) [inline silence/silence.go:628]:
```suggestion
			return "", fmt.Errorf("unable to query the silences while calculating the limit: %w", err)
```

File: silence/silence.go [hunk 6]
```diff
@@ -611,7 +646,11 @@ func (s *Silences) Set(sil *pb.Silence) (string, error) {
 		sil.StartsAt = now
 	}
 
-	return sil.Id, s.setSilence(sil, now, false)
+	if err = s.setSilence(sil, now, false); err != nil {
+		return "", err
+	}
+
+	return sil.Id, nil
```
Comment (gotjosh) [inline silence/silence.go:653]:
```suggestion
	return "", nil
```

Comment (gotjosh) [reply silence/silence.go:653]: Otherwise this doesn't match the comments.

File: silence/silence.go [hunk 7]
```diff
@@ -569,6 +580,13 @@ func (s *Silences) setSilence(sil *pb.Silence, now time.Time, skipValidate bool)
 		return err
 	}
 
+	// Check the limit unless the silence has been expired. This is to avoid
+	// situations where silences cannot be expired after the limit has been
+	// reduced.
+	if n := msil.Size(); s.limits.MaxPerSilenceSize > 0 && n > s.limits.MaxPerSilenceSize && sil.EndsAt.After(now) {
+		return fmt.Errorf("silence exceeded maximum size: %d", s.limits.MaxPerSilenceSize)
```
Comment (gotjosh) [inline silence/silence.go:587]:
```suggestion
		return fmt.Errorf("silence exceeded maximum size: %d bytes (limit: %d bytes)", msil.Size, s.limits.MaxPerSilenceSize)
```

Comment (gotjosh) [reply silence/silence.go:587]: This is the wording we use for other limits.

File: silence/silence_test.go [hunk 8]
```diff
@@ -458,6 +459,71 @@ func TestSilenceSet(t *testing.T) {
 	require.Equal(t, want, s.st, "unexpected state after silence creation")
 }
 
+func TestSilenceLimits(t *testing.T) {
```
Comment (gotjosh) [inline silence/silence_test.go:462]: Great test.

**Review by gotjosh — APPROVED** [review]
LGTM, but please take a look at my comments.

File: cmd/alertmanager/main.go [hunk 9]
```diff
@@ -146,6 +146,8 @@ func run() int {
 		dataDir             = kingpin.Flag("storage.path", "Base path for data storage.").Default("data/").String()
 		retention           = kingpin.Flag("data.retention", "How long to keep data for.").Default("120h").Duration()
 		maintenanceInterval = kingpin.Flag("data.maintenance-interval", "Interval between garbage collection and snapshotting to disk of the silences and the notification logs.").Default("15m").Duration()
+		maxSilences         = kingpin.Flag("silences.max-silences", "Maximum number of active and pending silences. If negative or zero, no limit is set.").Default("0").Int()
+		maxPerSilenceBytes  = kingpin.Flag("silences.max-per-silence-bytes", "Maximum per silence size. If negative or zero, no limit is set.").Default("0").Int()
```
Comment (gotjosh) [inline cmd/alertmanager/main.go:150]:
```suggestion
		maxPerSilenceBytes  = kingpin.Flag("silences.max-per-silence-bytes", "Maximum per silence size in bytes. If negative or zero, no limit is set.").Default("0").Int()
```

File: cmd/alertmanager/main.go [hunk 10]
```diff
@@ -146,6 +146,8 @@ func run() int {
 		dataDir             = kingpin.Flag("storage.path", "Base path for data storage.").Default("data/").String()
 		retention           = kingpin.Flag("data.retention", "How long to keep data for.").Default("120h").Duration()
 		maintenanceInterval = kingpin.Flag("data.maintenance-interval", "Interval between garbage collection and snapshotting to disk of the silences and the notification logs.").Default("15m").Duration()
+		maxSilences         = kingpin.Flag("silences.max-silences", "Maximum number of active and pending silences. If negative or zero, no limit is set.").Default("0").Int()
```
Comment (gotjosh) [inline cmd/alertmanager/main.go:149]:
```suggestion
		maxSilences         = kingpin.Flag("silences.max-silences", "Maximum number of active and pending silences excluding expired. If negative or zero, no limit is set.").Default("0").Int()
```

**Review by gotjosh — APPROVED** [review]
one last nit.

**Pull request merged**